gnovm/pkg/gnolang/gno_test.go

```diff
@@ -141,6 +141,31 @@ func TestRunEmptyMain(t *testing.T) {
 	m.RunMain()
 }
 
+func TestGCLoopy(t *testing.T) {
+	t.Parallel()
+	m := NewMachine("test", nil)
+	c := `package test
+
+func main() {
+	for i:=0;i<5000;i++ {
+		f := foo(i)
+	}
+}
+
+type Foo struct {
+	b int
+}
+
+func foo(b int) *Foo {
+	return &Foo{b: b}
+}
+`
+	n := MustParseFile("main.go", c)
+	m.RunFiles(n)
+	m.Alloc = NewAllocator(22000, 3000, NewHeap())
+	m.RunMain()
+}
+
 // run main() with a for loop.
```

Comment:
> I updated ShouldRunGC() so that we can control when to trigger the GC.
> 
> ```
> func (alloc *Allocator) ShouldRunGC() bool {
> 	// return false
> 	return alloc.bytes >= alloc.minGcRun
> }
> ```
> 
> For the following tests. regardless we triggered GC or not, Alloc.bytes are the same. Not sure if this is the expected behavior
> 
> run two simple GC deallocation test.
> 
> ```
> func TestGCAlloc(t *testing.T) {
> 	c := `package test
> func simpleGC() {
> 	{
> 		item1 := make([]byte, 100)
> 	}
> }`
> 	m := NewMachine("test", nil)
> 	n := MustParseFile("main.go", c)
> 	m.RunFiles(n)
> 	//m.Alloc = NewAllocator(22000, 10000, NewHeap())
> 	m.Alloc = NewAllocator(22000, 1, NewHeap())
> 	m.Eval(Call("simpleGC"))
> 	fmt.Printf("%+v\n", m.Alloc)
> 
> }
> 
> //&{maxBytes:22000 bytes:1140 lastGcRun:0 minGcRun:10000 heap:0x1400030bbc0 detonate:false}
> //&{maxBytes:22000 bytes:1140 lastGcRun:2348 minGcRun:1 heap:0x14000341bc0 detonate:false}
> ```
> 
> ```
> func TestGCAlloc(t *testing.T) {
> 	c := `package test
> func simpleGC() {
> 	{
> 		item1 := make([]byte, 100)
> 	}
> }`
> 	m := NewMachine("test", nil)
> 	n := MustParseFile("main.go", c)
> 	m.RunFiles(n)
> 	//m.Alloc = NewAllocator(22000, 10000, NewHeap())
> 	m.Alloc = NewAllocator(22000, 1, NewHeap())
> 	m.Eval(Call("simpleGC"))
> 	fmt.Printf("%+v\n", m.Alloc)
> 
> }
> 
> 
> // &{maxBytes:22000 bytes:1920 lastGcRun:2440 minGcRun:1 heap:0x14000330120 detonate:false}
> // &{maxBytes:22000 bytes:1920 lastGcRun:0 minGcRun:10000 heap:0x1400035a120 detonate:false}
> ```

Jae said we should only run the GC if we run out of memory. This is why it was commented and always returned false.

Reply:
The code you wrote does not reach the limit in memory so the GC is never triggered.

Reply:
Yes. These are expected because all the builtins and std are allocating too. So the allowed maximum allocation number needs to be increased.